ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala

```diff
@@ -394,12 +422,51 @@ class WebSocketService(
       request: A): Source[Error \/ Message, NotUsed] = {
     val Q = implicitly[StreamQuery[A]]
 
-    val (resolved, unresolved, fn) = Q.predicate(request, resolveTemplateId, lookupType)
+    val StreamPredicate(resolved, unresolved, fn, dbQuery) =
+      Q.predicate(request, resolveTemplateId, lookupType)
 
     if (resolved.nonEmpty) {
-      contractsService
-        .insertDeleteStepSource(jwt, parties, resolved.toList, offPrefix, Terminates.Never)
-        .via(convertFilterContracts(fn))
+      def prefilter: Future[(
+          Source[
+            ContractStreamStep[Nothing, (domain.ActiveContract[JsValue], Q.Positive)],
+            NotUsed],
+          Option[domain.StartingOffset])] =
+        contractsService.daoAndFetch
+          .filter(_ => offPrefix.isEmpty)
+          .cata(
+            {
+              case (dao, fetch) =>
+                val tx = for {
+                  bookmark <- fetch.fetchAndPersist(jwt, parties, resolved.toList)
+                  mdContracts <- dbQuery(parties)(dao.logHandler)
+                  // TODO SC: save bookmark here, loop if fail
+                } yield
+                  (
+                    Source.single(mdContracts).map(ContractStreamStep.Acs(_)) ++ Source.single(
+                      ContractStreamStep.LiveBegin(bookmark.map(_.toDomain))),
+                    bookmark.toOption map (term => domain.StartingOffset(term.toDomain)))
+                dao.transact(tx).unsafeToFuture()
+            },
+            Future.successful((Source.empty, offPrefix))
+          )
+
+      Source
```

Comment:
If I read this correctly we preserve the streaming behavior for the in-memory mode right? Just double-checking since I vaguely recall that at some point we discussed that this might also go away since (at least for now) we don’t stream db results.

Reply:
Yes; since we already had a separate ACS phase for the in-memory mode, we merely replace that "prefix".